pkg/local_object_storage/writecache/persist.go

```diff
@@ -79,15 +92,45 @@ func (c *cache) persistToCache(objs []objectInfo) []int {
 
 	var failDisk []int
 
-	for _, i := range failMem {
-		if uint64(len(objs[i].data)) > c.maxObjectSize {
-			failDisk = append(failDisk, i)
+	cacheSz := c.estimateCacheSize()
+
+	for _, objInd := range failMem {
+		var (
+			updCacheSz  uint64
+			overflowInd = -1
+		)
+
+		if objInd < 0 {
+			// actually, since the overflow was detected in DB tx, the required space could well have been freed,
+			// but it is easier to consider the entire method atomic
+			overflowInd = -objInd
+		} else {
+			// check if object will overflow write-cache size limit
+			if updCacheSz = c.incSizeFS(cacheSz); updCacheSz > c.maxCacheSize {
+				overflowInd = objInd
+			}
+		}
+
+		if overflowInd >= 0 {
+			for j := range objs[overflowInd:] {
```

Comment:
If I understand correctly, `objs` will contain both `doneMem` and `failMem` objects at this point of execution. In this case with `range objs[overflowInd:]`, you can mark some objects from `doneMem` slice. 

Example: consider `objs [0, 1, 2, 3]` where `1 2` --  big objects, `0 3` -- small objects. bbolt.DB is exhausted. Then `failMem` contains [-1,-2]. Therefore objs[1:] will be added to `failDisk` including 3, which might be incorrect.

Reply:
Yes, we need to exclude indices from `doneMem`, fixed.